File: 210-course-schedule-ii/210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int nc, int[][] pre) {
        
        Map<Integer, Integer> dep = new HashMap<>();
        Map<Integer, Set<Integer>> req = new HashMap<>();
        
        for(int[] p : pre) {
            Set<Integer> s = req.computeIfAbsent(p[1], i -> new HashSet<>());
            s.add(p[0]);
            dep.put(p[0], dep.getOrDefault(p[0], 0) + 1);
        }
        
        Queue<Integer> q = new LinkedList<>();
        List<Integer> res = new ArrayList<>();
        
        for(int i = 0; i < nc; i++) {
            if (dep.containsKey(i)) continue;
            q.offer(i);
        }
        
        while(!q.isEmpty()) {
            Integer i = q.poll();
            res.add(i);
            Set<Integer> s = req.get(i);
            if (s == null) continue;
            for(Integer si : s) {
                Integer cnt = dep.get(si);
                if (cnt != 1) {
                    dep.put(si, cnt - 1);
                    continue;
                }
                dep.remove(si);
                q.offer(si);
            }
        }
        
        if (!dep.isEmpty()) return new int[0];
        
        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: 210-course-schedule-ii/210-course-schedule-ii.scala
import scala.collection.mutable._

object Solution {
    def findOrder(numCourses: Int, prerequisites: Array[Array[Int]]): Array[Int] = {
        val prereq: Array[Set[Int]] = Array.fill(numCourses)(Set())
        val reqBy: Array[Set[Int]] = Array.fill(numCourses)(Set())

        prerequisites.foreach { pre => 
            prereq(pre(0)).add(pre(1)) 
            reqBy(pre(1)).add(pre(0))
        }

        val queue: Queue[Int] = Queue()

        prereq.zipWithIndex.foreach { case (req, i) => if (req.isEmpty) queue.enqueue(i) }

        val res: ListBuffer[Int] = ListBuffer()

        while(queue.nonEmpty) {
            val n = queue.dequeue
            res += n
            reqBy(n).foreach { c =>
                prereq(c).remove(n)
                if (prereq(c).isEmpty) queue.enqueue(c)
            }
        }

        if (res.size == numCourses) res.toArray else Array[Int]()
    }
}